Rebuild state

tiagolobocastro opened the pull request:
rebuild: split rebuild module

splits the rebuild module into impl and api
no functional changes
this should help contain the next set of coming changes

-----------------------------------------------------

rebuild: split the rebuild state logic

Fixes the rebuild state by introducing and current and a pending state
the pending state is set by the nexus whenever it wants to perform
an operation on the rebuild. The pending state is reconciled when the
job's runner adjusts to the pending state, eg running -> paused

The states are now fully validated by a state machine. No catch all
"_ => ()", was used so the compiler should complain if a new state is
added.

If a pending state is "pending" then a new operation is rejected until
the previous completes.
Internal operations are allowed to override the pending state

Started to make the trait a first class citizen

bonus: fully implement the child src removal by waiting on the rebuild
job to complete before creating a new job to replace the removed one.
This means we now enable the commented out test

tiagolobocastro committed:
rebuild: Move rebuild task list from the nexus into a global scope (#149)

This makes it easier to interact with the rebuild task list using its destination alone as a key
Also tidy up some of the doc comments and reorder the code a bit

also bring back CI yaml file since we've rebased master off of openebs/master

Commit:
Add pause/resume rebuild support (#158)

tiagolobocastro committed:
Rename RebuildTask to RebuildJob and move rebuild code...

... to a separate file to facilitate mainline rebases
(plus we can always change it back when we integrate into master)

tiagolobocastro committed:
rebuild: run n async copy tasks at a time (#162)

a management async tasks schedules all copy tasks, awaits them and 
schedules enough tasks until the bdev is fully rebuilt
(still using a single reactor this time)

Commit:
Rebuild: Start rebuild when setting child online (#164)

Start a rebuild when setting child online.

- When a child is set offline, stop any ongoing rebuilds for that child
- When setting a child online, set its state to faulted and initiate a rebuild
- Add support for offline and online operations to the "child_operation"

tiagolobocastro committed:
rebuild: remove the teamrebuild CI

remove the test yml as it is not required anymore

tiagolobocastro committed:
rebuild: clean up rebuild job on child removal

when a child is being remove we check for any existing jobs that has:
1. child as dest
in this case we simply stop the job and wait for it to finish
2. child as src
same as above but we also need to start a new rebuild job to rebuild the
dest child.

Atm the wait infra is not here so we have to rely on the child destroy
to block until all the bdevs are closed.
This will be addressed once cas-194 is resolved.

CAS-126

tiagolobocastro committed:
rebuild: add missing_docs attribute

various bits of the rebuild have now been documented
any missing doc will now be a compiler warning

CAS-201

tiagolobocastro committed:
move rebuild and replica into replica dir

tiagolobocastro committed:
rebuild: split rebuild module

splits the rebuild module into impl and api
no functional changes
this should help contain the next set of coming changes
todo: completely hide the rebuild job and expose the trait only

tiagolobocastro committed:
rebuild: split the rebuild state logic

Fixes the rebuild state by introducing and current and a pending state
the pending state is set by the nexus whenever it wants to perform
an operation on the rebuild. The pending state is reconciled when the
job's runner adjusts to the pending state, eg running -> paused

The states are now fully validated by a state machine. No catch all
"_ => ()", was used so the compiler should complain if a new state is
added.

If a pending state is "pending" then a new operation is rejected until
the previous completes.
Internal operations are allowed to override the pending state

Started to make the trait a first class citizen

bonus: fully implement the child src removal by waiting on the rebuild
job to complete before creating a new job to replace the removed one.
This means we now enable the commented out test

CAS-194

paulyoong commented on mayastor/src/bdev/nexus/nexus_bdev_rebuild.rs:
```diff
@@ -61,24 +61,39 @@ impl Nexus {
+
+    /// Terminates a rebuild in the background
+    /// used for shutdown operations and
+    /// unlike the client operation stop, this command does not fail
+    /// as it skips previous client operations
```
Is it accurate to say it skips previous client operations? They may already be in progress. Is it more of a case of overwriting the desired state.

tiagolobocastro replied:
skips -> overrides ?

paulyoong replied:
I think so

tiagolobocastro replied:
Done

paulyoong commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
+}
+
+impl Default for RebuildStates {
+    fn default() -> Self {
+        Self {
+            current: RebuildState::Init,
+            pending: None,
+        }
+    }
+}
```
Given that these are being set to their default values, could we use Default::default()?

tiagolobocastro replied:
Done

paulyoong commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
+            }
+            RebuildOperation::Pause => match self.state() {
+                S::Stopped | S::Failed | S::Completed => Err(e),
+                S::Init | S::Running | S::Paused => {
```
Should the paused state be split out on its own so it resembles the other cases i.e. "for idempotence sake" on line 584?

tiagolobocastro replied:
Done

paulyoong commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
+            },
+            RebuildOperation::Fail => match self.state() {
+                S::Init | S::Stopped | S::Paused | S::Completed => Err(e),
+                S::Running | S::Failed => {
```
Again, should the failed state be split out on its own?

tiagolobocastro replied:
Done

paulyoong commented on mayastor/src/replicas/rebuild/rebuild_impl.rs:
```diff
@@ -585,3 +462,175 @@ impl RebuildJob {
+                }
+            },
+            RebuildOperation::Complete => match self.state() {
+                S::Init | S::Paused | S::Stopped | S::Failed | S::Completed => {
```
If its already completed should it just return Ok?

tiagolobocastro replied:
hmm I was trying to signal that you shouldn't have called complete twice as it only completes when you "complete"

paulyoong replied:
Ok

tiagolobocastro committed:
Address review comments, fix js test and...

...also fix issues pointed out by semi-standard
reduce the wait time we wait for the stop as it speeds up the
js test cycles
